src/main/java/org/folio/services/invoice/InvoiceCancelService.java

```diff
@@ -224,4 +239,116 @@ private Future<Void> unreleaseEncumbrancesForPoLines(List<PoLine> poLines, Invoi
+    return "paymentStatus==(\"Fully Paid\" OR \"Partially Paid\") AND " + convertIdsToCqlQuery(poLineIds);
+  }
+
+  private String queryToGetRelatedPaidInvoiceLinesByPoLineIds(String invoiceId, List<String> poLineIds) {
+    return "invoiceId<>" + invoiceId + " AND invoiceLineStatus==(\"Paid\") AND " +
+      convertIdsToCqlQuery(poLineIds, "poLineId", true);
+  }
+
+  private Future<List<PoLine>> getPoLinesByIdAndQuery(List<String> poLineIds, Function<List<String>, String> queryFunction,
+      RequestContext requestContext) {
+    List<Future<List<PoLine>>> futureList = StreamEx
+      .ofSubLists(poLineIds, MAX_IDS_FOR_GET_RQ)
+      .map(queryFunction)
+      .map(query -> orderLineService.getPoLines(query, requestContext))
+      .toList();
+
+    return collectResultsOnSuccess(futureList)
+      .map(col -> col.stream().flatMap(List::stream).toList());
+  }
+
+  private Future<List<InvoiceLine>> getInvoiceLinesByPoLineIdsAndQuery(List<String> poLineIds,
+      Function<List<String>, String> queryFunction, RequestContext requestContext) {
+    List<Future<List<InvoiceLine>>> futureList = StreamEx
+      .ofSubLists(poLineIds, MAX_IDS_FOR_GET_RQ)
+      .map(queryFunction)
+      .map(query -> invoiceLineService.getInvoiceLinesByQuery(query, requestContext))
+      .toList();
+
+    return collectResultsOnSuccess(futureList)
+      .map(col -> col.stream().flatMap(List::stream).toList());
+  }
+
+  private Future<List<Invoice>> getInvoicesByIdsAndQuery(List<String> invoiceIds, String query, RequestContext requestContext) {
+    String query2 = "(" + query + ") AND " + convertIdsToCqlQuery(invoiceIds);
+    return invoiceService.getInvoices(query2, 0, Integer.MAX_VALUE, requestContext)
+      .map(InvoiceCollection::getInvoices);
+  }
+
+  private Future<Void> updateRelatedPoLines(List<InvoiceLine> allRelatedInvoiceLines, List<PoLine> filteredPoLines,
+      List<Invoice> relatedInvoices, RequestContext requestContext) {
+    List<String> relatedInvoiceIds = relatedInvoices.stream().map(Invoice::getId).toList();
+    List<InvoiceLine> filteredInvoiceLines = allRelatedInvoiceLines.stream()
+      .filter(invoiceLine -> relatedInvoiceIds.contains(invoiceLine.getInvoiceId()))
+      .toList();
+    Map<String, List<InvoiceLine>> poLineIdToInvoiceLines = filteredInvoiceLines.stream()
+      .collect(groupingBy(InvoiceLine::getPoLineId));
+    List<PoLine> modifiedPoLines = new ArrayList<>();
+    filteredPoLines.forEach(poLine -> {
+      List<InvoiceLine> relatedInvoiceLines = poLineIdToInvoiceLines.get(poLine.getId());
+      if (relatedInvoiceLines == null || relatedInvoiceLines.isEmpty()) {
+        poLine.setPaymentStatus(AWAITING_PAYMENT);
+        modifiedPoLines.add(poLine);
+      } else if (relatedInvoiceLines.stream().noneMatch(InvoiceLine::getReleaseEncumbrance) &&
+          !PARTIALLY_PAID.equals(poLine.getPaymentStatus())) {
+        poLine.setPaymentStatus(PARTIALLY_PAID);
+        modifiedPoLines.add(poLine);
+      }
```

Comment:
I am not sure what you mean. The second part is "`else if`", not "`else`". If I just add `modifiedPoLines.add(poLine);` after the `if` it will not have the same logic. And I prefer to repeat the `add()` rather than the conditions.

Reply:
ok